### PR TITLE
Simple batch normalization that doesn't update the ema mean and var during test

### PR DESCRIPTION
**Pull request by BigRedT**
Added:
- `BatchNorm` class 
- `test_batchnorm_train_mode()` 
- `test_batchnorm_test_mode()`

Help needed !!
This script works if used with the default graph as shown in the unit tests.
Currently this doesn't work if I try to scope it under a specific graph as follows
```
g = tf.Graph()
with g.as_default():
    bn = BatchNorm(x, training, name='bn')
    y = bn.output
```
The error I get says that self.mean is not in the same graph as Exponential Moving Average and I verified that indeed `self.mean.graph==g` evaluates to `False` (having passed `g` into the `BatchNorm` class). This error is thrown in line `ema_apply_op = ema.apply([self.mean, self.variance])`

### CODE REVIEW

**Comment by kevjshih** [comment]
I'm a little confused as to what's going on here. Why is the batchnorm implementation now twice as long with a helper class?

**Comment by BigRedT** [comment]
@kevjshih BatchNormalizer was what @jrock08 had written before. I didn't change it (yapf probably did some indentation).

My implementation (taken from Saurabh and modified slightly) is class BatchNorm which is simpler both in terms of code and usage than Jason's. I think this would replace BatchNormalizer for most purposes if not all. So you should only review that class implementation and the corresponding unit tests - test_batchnorm_train_mode() and test_batchnorm_test_mode()

**Comment by kevjshih** [comment]
Gotcha, so once it works, we'll replace Jason's if it's simpler? That makes sense. I'll take a look at the test case it's complaining about maybe tomorrow.

**Comment by BigRedT** [comment]
Yeah. @jrock08 said he has a special use case where his version might still be needed but for most of us the simpler version should be more than sufficient.

**Comment by kevjshih** [comment]
Which special case? I guess in terms of usage, both implementations are equivalent in simplicity.

**Comment by kevjshih** [comment]
I guess in terms of code length, it's really just 10 or so lines shorter. I don't think we should keep two versions around if both work and are equally simple to use.

**Comment by BigRedT** [comment]
Yeah I am not sure what his use case was. I will let him answer that. If both work in all cases then the second one is strictly easier to understand and I think also easier to use (no wrapper required around the original class like `batch_normalize`)

**Comment by jrock08** [comment]
Duplicating a graph with batch norm layers (eg. adversarial stuff or siamese stuff)

**Comment by kevjshih** [comment]
If the only difference is the loss of the weight reuse functionality, I'm not sure there's much reason to replace the current implementation. Aren't both of them exactly one function call at most? The wrapper shouldn't matter to the user since it's not part of the exposed interface anyways.

**Comment by jrock08** [comment]
This implementation correctly updates the EMA without external calls which is very nice.  I'm planning on trying to come up with a reuse solution after this goes in and remove the duplicate code.

**Comment by kevjshih** [comment]
Ok, that makes sense. To do that for this implementation, would it work to just move lines 120-132 into the init function. That way, the graph variables don't reinitialized when just calling BatchNorm.batchnorm()

**Comment by jrock08** [comment]
I think we can close this as duplicate of #28

**Comment by jrock08** [comment]
I think we can close this as duplicate of #28
